File: 16) Maximum Product of Word Lengths/prod.py
def maxProduct(words):
    max_product = 0
    word_bits = [0] * len(words)
    
    for i, word in enumerate(words):
        for char in word:
            word_bits[i] |= 1 << (ord(char) - ord('a'))
    
    for i in range(len(words)):
        for j in range(i+1, len(words)):
            if word_bits[i] & word_bits[j] == 0:
                max_product = max(max_product, len(words[i]) * len(words[j]))
    
    return max_product

words = list(map(str, input("Enter strings: ").split()))
print(maxProduct(words)) 

File: 16) Maximum Product of Word Lengths/prod.java
import java.util.*;

public class Main {
    public static int maxProduct(String[] words) {
        int maxProduct = 0;
        int[] wordBits = new int[words.length];
        
        for (int i = 0; i < words.length; i++) {
            for (char c : words[i].toCharArray()) {
                wordBits[i] |= 1 << (c - 'a');
            }
        }
        
        for (int i = 0; i < words.length; i++) {
            for (int j = i + 1; j < words.length; j++) {
                if ((wordBits[i] & wordBits[j]) == 0) {
                    maxProduct = Math.max(maxProduct, words[i].length() * words[j].length());
                }
            }
        }
        
        return maxProduct;
    }
    
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.print("Enter words: ");
        String input = sc.nextLine();
        String[] words = input.split(" ");
        System.out.println(maxProduct(words));

        sc.close();
    }
}
